### PR TITLE
Don't traverse directories that match file_ignore_regex and file_ignore_glob

### LINKED ISSUE #33907: Fileserver traverses directories that match file_ignore_regex (highstate slow over vagrant share)
### Description of Issue/Question

When running a salt masterless highstate, we see that the fileserver calls os.walk on the roots dir, which walks all directories in the file roots. This includes directories that we do not want to serve, and that we do not want added to the fileserver cache for performance reasons. For example, if we have "^\.git" in our file_include_regex, we do not want the fileserver to go into our .git directories and add all the git internal files to its fileserver cache.

Is it sufficient to think that the file_ignore_regex should cover this? Or should a new dir_ignore_regex / dir_ignore_glob config option be created instead?

### Background

In our setup our salt states repos live on a host machine and the salt-master is running in a VM. It accesses the salt states over a vagrant share. Because the fileserver walks the entire roots directory, it ends up calling stat() on thousands of files in our .git directories that we want ignored. Since we have multiple git directories in the file roots, this amounts to about 10000+ stat() calls every time salt fileserver cache gets updated (which is many times during a masterless highstate). Since this is over a vagrant share, the stat() calls have much higher latency than stat() on a local file system.

Sample file roots on host machine:
```
/srv/salt/.git
/srv/salt/top.sls
/srv/salt/stacks/app1-stack/.git
/srv/salt/stacks/app1-stack/init.sls
/srv/salt/stacks/app2-stack/.git
/srv/salt/stacks/app2-stack/init.sls
...
...
```

Our file_roots and file_ignore_regex:
```
file_roots:
  base:
    - /srv/salt

file_ignore_regex:
  - '^\.git($|/)'
``` 

### Proposal

I was able to improve highstate performance over a Vagrant share from 15+ minutes to ~2 minutes by removing directories that fileserver walks if it matches file_ignore_regex.

The following sample change was made to various parts of the code that call os.walk() (mainly in salt/fileserver/roots.py). It works by removing directories that os.walk() will go into before it traverses them. (https://docs.python.org/2/library/os.html "... When topdown is True, the caller can modify the dirnames list in-place (perhaps using del or slice assignment), and walk() will only recurse into the subdirectories whose names remain in dirnames; this can be used to prune the search)

```
    for root, dirs, files in os.walk(
            path,
            followlinks=__opts__['fileserver_followsymlinks']):
+        dirs[:] = [d for d in dirs if not salt.fileserver.is_file_ignored(__opts__, d)]
```

I can submit a PR with all the full changes if this is something we can move forward with, or if we should have a separate dir_ignore_regex, dir_ignore_glob config options, and is_dir_ignored() functions.

### Versions Report
Salt Version:
           Salt: 2016.3.0

Dependency Versions:
           cffi: Not Installed
       cherrypy: 3.5.0
       dateutil: 2.4.2
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.11+ (default, Apr 17 2016, 14:00:29)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: Ubuntu 16.04 xenial
        machine: x86_64
        release: 4.4.0-21-generic
         system: Linux
        version: Ubuntu 16.04 xenial

### PR DESCRIPTION
**Pull request by ctrlrsf**
### What does this PR do?

This PR allows salt to ignore directories that match regex or globs in file_ignore_regex or file_ignore_glob settings. This is helpful if you have a large directory you want ignored so that salt does not waste time learning about files in that directory. One such example is the .git directory which can be very large. When file roots is on a network drive or vagrant share, learning of all these files severely degrades salt-master and salt-call state.highstate performance due to os.stat calls that are called on each file. (benchmarks below)
### What issues does this PR fix or reference?

This PR fixes https://github.com/saltstack/salt/issues/33907
### Previous Behavior

Salt roots fileserver periodically generates a list/cache of files that it can serve. Doing so ends up calling os.stat on each file, which is can be slow over a network or vagrant share. If a file matches file_ignore_regex or file_ignore_glob, then the file is excluded from the list. If the directory name matches file_ignore_regex or glob, salt still goes into directory and adds all those files to its file cache.
### New Behavior

As salt generates file list, if it encounters a directory whose name matches file_ignore_regex or glob, it is skipped and all files within that directory are ignored and not cached.
### Benchmarks

Two benchmarks below show state.highstate now completes in 2 minutes instead of 10 minutes when I ignore the .git directory and file roots is on a vagrant share. strace -c shows the reduction in stat syscalls from 582314 calls  to 140998. Both of these highstates were done after an initial highstate was already completed, so all states with side-effects had already been completed.

strace -c of salt-call state.highstate and file_ignore_regex contain '.git':

```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 74.87    8.855751          63    140998      3708 stat
 13.25    1.567275          38     40854     15411 open
 10.66    1.261491          40     31720        13 lstat
  0.47    0.055037           2     30559           fstat
  0.43    0.051170         391       131           clone
  0.12    0.014648           0     39136           getdents
  0.08    0.009545           0     26064           close
  0.07    0.008091          62       131           wait4
  0.02    0.002637           0      9225         2 read
  0.02    0.002008           0      4695           write
  0.00    0.000564           0      5333       935 lseek
  0.00    0.000135           1       111           poll
  0.00    0.000114           1       224           umask
  0.00    0.000111           0      5515           fcntl
  0.00    0.000099           0      2225           mmap
  0.00    0.000082           0      1794           munmap
  0.00    0.000055           1        55           setresgid
  0.00    0.000052           0       116           getcwd
  0.00    0.000000           0       252           mprotect
  0.00    0.000000           0       228           brk
  0.00    0.000000           0        69           rt_sigaction
  0.00    0.000000           0         3           rt_sigprocmask
  0.00    0.000000           0         9         2 ioctl
  0.00    0.000000           0       922       694 access
  0.00    0.000000           0       295           pipe
  0.00    0.000000           0         1           msync
  0.00    0.000000           0         3           dup
  0.00    0.000000           0        21           socket
  0.00    0.000000           0        21        19 connect
  0.00    0.000000           0         2           sendto
  0.00    0.000000           0         2           recvfrom
  0.00    0.000000           0         1           execve
  0.00    0.000000           0         1           kill
  0.00    0.000000           0        16           uname
  0.00    0.000000           0       474           flock
  0.00    0.000000           0         1           rename
  0.00    0.000000           0        18           mkdir
  0.00    0.000000           0        15           rmdir
  0.00    0.000000           0        64         1 unlink
  0.00    0.000000           0        13         2 readlink
  0.00    0.000000           0         1           chmod
  0.00    0.000000           0         1           chown
  0.00    0.000000           0         2           getrlimit
  0.00    0.000000           0         1           getrusage
  0.00    0.000000           0         3           sysinfo
  0.00    0.000000           0         3           getuid
  0.00    0.000000           0         2           getgid
  0.00    0.000000           0         2           geteuid
  0.00    0.000000           0        37           getegid
  0.00    0.000000           0         1         1 statfs
  0.00    0.000000           0         1           arch_prctl
  0.00    0.000000           0         1           gettid
  0.00    0.000000           0        10           futex
  0.00    0.000000           0         1           set_tid_address
  0.00    0.000000           0         1           set_robust_list
  0.00    0.000000           0         4           pipe2
------ ----------- ----------- --------- --------- ----------------
100.00   11.828865                341388     20788 total

real    2m37.927s
user    0m16.064s
sys     1m6.400s
```

timed strace -c of salt-call state.highstate (not ignoring .git directory):

```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 70.21   30.634259          53    582314      3693 stat
 18.42    8.035836          68    118804     15410 open
 10.60    4.625255          39    118224        13 lstat
  0.47    0.205072           2    108526           fstat
  0.16    0.069499           0    195026           getdents
  0.10    0.042764           0    104015           close
  0.01    0.006235           1      8789           write
  0.01    0.006036          46       131           wait4
  0.01    0.003156          24       131           clone
  0.00    0.001183           0     11602         2 read
  0.00    0.000436           0      2060           munmap
  0.00    0.000392           0      2506           mmap
  0.00    0.000165           0      5529           fcntl
  0.00    0.000161           0      5357       935 lseek
  0.00    0.000112           0       915       687 access
  0.00    0.000000           0       111           poll
  0.00    0.000000           0       252           mprotect
  0.00    0.000000           0       230           brk
  0.00    0.000000           0        69           rt_sigaction
  0.00    0.000000           0         3           rt_sigprocmask
  0.00    0.000000           0         9         2 ioctl
  0.00    0.000000           0       295           pipe
  0.00    0.000000           0         1           msync
  0.00    0.000000           0         3           dup
  0.00    0.000000           0        21           socket
  0.00    0.000000           0        21        19 connect
  0.00    0.000000           0         2           sendto
  0.00    0.000000           0         2           recvfrom
  0.00    0.000000           0         1           execve
  0.00    0.000000           0         1           kill
  0.00    0.000000           0        16           uname
  0.00    0.000000           0       476           flock
  0.00    0.000000           0       118           getcwd
  0.00    0.000000           0        13           mkdir
  0.00    0.000000           0        13           rmdir
  0.00    0.000000           0        64         1 unlink
  0.00    0.000000           0        13         2 readlink
  0.00    0.000000           0       224           umask
  0.00    0.000000           0         2           getrlimit
  0.00    0.000000           0         1           getrusage
  0.00    0.000000           0         3           sysinfo
  0.00    0.000000           0         3           getuid
  0.00    0.000000           0         2           getgid
  0.00    0.000000           0         2           geteuid
  0.00    0.000000           0        37           getegid
  0.00    0.000000           0        55           setresgid
  0.00    0.000000           0         1         1 statfs
  0.00    0.000000           0         1           arch_prctl
  0.00    0.000000           0         1           gettid
  0.00    0.000000           0        10           futex
  0.00    0.000000           0         1           set_tid_address
  0.00    0.000000           0         1           set_robust_list
  0.00    0.000000           0         4           pipe2
------ ----------- ----------- --------- --------- ----------------
100.00   43.630561               1266011     20765 total

real    10m28.216s
user    0m29.580s
sys     4m41.004s
```

### CODE REVIEW

**Comment by cachedout** [comment]
This looks good to me. I would also like @terminalmage to have a look.

File: salt/fileserver/__init__.py [hunk 1]
```diff
@@ -171,14 +171,16 @@ def check_env_cache(opts, env_cache):
...
+def generate_mtime_map(opts, path_map):
     '''
     Generate a dict of filename -> mtime
     '''
     file_map = {}
     for saltenv, path_list in six.iteritems(path_map):
         for path in path_list:
             for directory, dirnames, filenames in os.walk(path):
+                # Don't walk any directories that match file_ignore_regex or glob
+                dirnames[:] = [d for d in dirnames if not is_file_ignored(opts, d)]
```
Comment (terminalmage) [inline salt/fileserver/__init__.py]: Is there any reason (outside of not allocating a new list in memory) for using slice notation here?

**Comment by cachedout** [comment]
@ctrlrsf Could you please have a look at the comments from @terminalmage ?

**Comment by ctrlrsf** [comment]
Sorry for the delay, was traveling. Slice assignment was used to modify dirnames in place. From os.walk documentation:

https://docs.python.org/2/library/os.html#os.walk
> ... the caller can modify the dirnames list in-place (perhaps using del or slice assignment), and walk() will only recurse into the subdirectories whose names remain in dirnames; this can be used to prune the search

I thought using a slice assignment would be the most straight forward way to do so since its mentioned directly in the os.walk docs. If you prefer this be done another way, please let me know though.

**Comment by terminalmage** [comment]
@ctrlrsf Awesome, I had never come across this before. Thanks for clarifying, and helping me to learn something new today.

**Pull request merged**